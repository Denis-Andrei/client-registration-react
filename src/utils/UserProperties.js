export let UserProperties =  {
    crn: "crn",
    name: "name",
    businessName: "businessName",
    contactNumber: "contactNumber",
    propertyNumber: "propertyNumber",
    postcode: "postcode",
    businessType: "businessType",
    password: "password",
    passwordCheck: "passwordCheck"
}

