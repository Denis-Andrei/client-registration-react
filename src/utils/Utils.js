


export const getData = (name) => sessionStorage.getItem(name) ? sessionStorage.getItem(name) : ""

export const isNameValid = (name) => /^([A-z])+(.?[a-zA-Z])*('?[a-zA-Z])*/.test(name)
export const isNumberValid = (number) => /^[0-9]{10}$|^[0-9]{11}$/.test(number) 
export const isPostcodeValid = (postcode) => /^(?:[A-Za-z]\d ?\d[A-Za-z]{2})|(?:[A-Za-z][A-Za-z\d]\d ?\d[A-Za-z]{2})|(?:[A-Za-z]{2}\d{2} ?\d[A-Za-z]{2})|(?:[A-Za-z]\d[A-Za-z] ?\d[A-Za-z]{2})|(?:[A-Za-z]{2}\d[A-Za-z] ?\d[A-Za-z]{2})/.test(postcode) 
export const isPasswordValid = (password) => password.length >= 10 
export const isPasswordMatching = (password, checkPassword) => password === checkPassword

