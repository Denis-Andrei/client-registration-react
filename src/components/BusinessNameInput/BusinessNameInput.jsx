import React, {useState} from 'react'
import { BackLink, H1, Label, LabelText, HintText, ErrorText, Input, FormGroup } from 'govuk-react'
import { Button } from 'govuk-react'
import { useHistory } from 'react-router-dom';
import { getData, isNameValid } from '../../utils/Utils';
import { UserProperties } from '../../utils/UserProperties';



export default function BusinessNameInput() {
    const history = useHistory();

    const [businessName, setBusinessName] = useState(getData(UserProperties.businessName));
    const [hasErrors, setHasErrors] = useState(false)

    const handleChange = (e) => {        
        setBusinessName(e.target.value)
    }

    const handleSubmit = () => {
        if(isNameValid(businessName)){ 
            sessionStorage.setItem(UserProperties.businessName, businessName);
            history.push('/contact-number')
        }else{
            setHasErrors(true)
        }
    }
     
    return (
        <div>
            <BackLink data-testid="goBack" onClick={ () => history.goBack()}>Back</BackLink>
            <H1 data-testid="heading">Enter Business Name</H1>
            <FormGroup>
                <Label>
                    <LabelText data-testid="label">Business Name</LabelText>
                    <HintText data-testid="hint">Enter the name of your business, this cannot be changed once submitted at the end, for example - 'McDonalds'</HintText>
                    {
                        hasErrors 
                        ? 
                        <ErrorText>Please provide your businesses name</ErrorText>
                        : 
                        ""
                    }
                    <Input data-testid="input-business-name" value={businessName} onChange={handleChange} required/>
                </Label>
            </FormGroup>
            <Button data-testid="submit" onClick={handleSubmit}>Continue</Button>
        </div>
    )
}
