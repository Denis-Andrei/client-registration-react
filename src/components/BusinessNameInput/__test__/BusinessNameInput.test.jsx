import React from 'react';
import ReactDOM from 'react-dom';
import BusinessNameInput from '../BusinessNameInput';
import { render, cleanup, fireEvent } from '@testing-library/react'
import renderer from "react-test-renderer"
import * as Utils from '../../../utils/Utils'

afterEach(cleanup);

//Mocking history.push() and history.goBack()
const mockHistoryPush = jest.fn();
const mockGoBack = jest.fn();

    jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockHistoryPush,
        goBack: mockGoBack
    }),
}));
//end of mocking


it("BusinessNameInput() renders withoud crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<BusinessNameInput />, div) 
})

it("renders BusinessNameInput() correctly", () => {
    const { getByTestId }  = render(<BusinessNameInput />)
    expect(getByTestId('heading')).toHaveTextContent("Enter Business Name")
    expect(getByTestId('label')).toHaveTextContent("Business Name")
    expect(getByTestId('hint')).toHaveTextContent("Enter the name of your business, this cannot be changed once submitted at the end, for example - 'McDonalds'")
})

it("matches snapshot", () => {
    const tree = renderer.create(<BusinessNameInput />).toJSON();
    expect(tree).toMatchSnapshot();
})

describe("handleChange() ", () => {
    it("updates businessName input on change event", () => {
        const { getByTestId } = render(<BusinessNameInput />)
        const businessNameInput = getByTestId("input-business-name");
    
        fireEvent.change(businessNameInput, {target: {value: "test"}})
        expect(businessNameInput.value).toBe("test")
    })
})

describe("handleSubmit() ", () => {
    it("onSubmit redirects to the correct route", () => {
        
        const mockIsNameValid =  jest.fn().mockReturnValue(true)
        Utils.isNameValid = mockIsNameValid

        const { getByTestId } = render(<BusinessNameInput />)
        const submit = getByTestId("submit");

        fireEvent.click(submit);
        expect(mockHistoryPush).toHaveBeenCalledWith('/contact-number');
    })
})

describe("goBack() ", () => {
    it("should be called on click event", () => {
        const { getByTestId } = render(<BusinessNameInput />)
        const goBackBtn = getByTestId("goBack");
    
        fireEvent.click(goBackBtn)
        expect(mockGoBack).toHaveBeenCalled();
    })
})