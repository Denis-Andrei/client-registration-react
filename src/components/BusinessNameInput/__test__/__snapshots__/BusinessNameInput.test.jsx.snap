// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <a
    className="src__Anchor-sc-11edswm-0 gGueb"
    data-testid="goBack"
    onClick={[Function]}
  >
    Back
  </a>
  <h1
    className="heading__StyledHeading-sc-1sovo83-0 juamsS"
    data-testid="heading"
    size="XLARGE"
  >
    Enter Business Name
  </h1>
  <div
    className="src__FormGroup-kqurat-0 fREKLF"
  >
    <label
      className="src__StyledLabel-sc-14n2scx-0 kHdLXy"
    >
      <span
        className="src__StyledLabelText-sc-12l0n4z-0 HDdRl"
        data-testid="label"
      >
        Business Name
      </span>
      <span
        className="src__StyledHint-sc-1qn4br2-0 kngyqG"
        data-testid="hint"
      >
        Enter the name of your business, this cannot be changed once submitted at the end, for example - 'McDonalds'
      </span>
      
      <input
        className="src__StyledInput-u8exb4-0 iqwpbG"
        data-testid="input-business-name"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </label>
  </div>
  <button
    className="sc-bdnxRM gXVtdA"
    data-testid="submit"
    disabled={false}
    onClick={[Function]}
  >
    Continue
  </button>
</div>
`;
